File: 1.BubbleSort.py

'''
bubble sort implementation using a for loop

Big-O
    Time Complexity  :
    Space Complexity :
'''

unsorted_listay = [64,34,25,12,22,11,90,5]

def bubble_sort(list, order = "ascending"):
    n = len(list)

    "ascending order"
    def asc():
        if list[j] > list[j + 1]:
                list[j + 1], list[j] = list[j], list[j + 1]

    def desc():
         if list[j] < list[j + 1]:
                list[j + 1], list[j] = list[j], list[j + 1]

    for i in range(n - 1) :
        for j in range(n - 1 - i) :
            if order == "descending":
                 desc()
            else:
                asc()

    return list

print(bubble_sort(unsorted_listay, "ascending"))

File: 1.BubbleSort.ts
/**
 * bubble sort in typescript
 * space complexity : constant O(1)
 * time complexity  : quadratic O(n^2)
 */

const unsortedArray: number[] = [64, 34, 25, 12, 22, 11, 90, 5];

function bubbleSort(arr: number[] = [], order: string = "ascending"): number[] {
  const n = arr.length;

  for (let i = 0; i < n - 1; i++) {
    for (let j = 0; j < n - 1 - i; j++) {
      if (
        (order === "descending" && arr[j] < arr[j + 1]) ||
        (order === "ascending" && arr[j] > arr[j + 1])
      ) {
        [arr[j], arr[j + 1]] = [arr[j + 1], arr[j]]; // Swap elements
      }
    }
  }

  return arr;
}

console.log(bubbleSort(unsortedArray));
